MRG: deprecate ddof / bias args to corrcoef

matthew-brett opened the pull request:
As discussed on the mailing list, the values for the bias and ddof
arguments to the corrcoef cancel in calculation, and are therefore
pointless and confusing.

Deprecate these arguments and document their pending removal.

This PR also includes a version of the warnings.catch_warnings decorator
that allows warnings to be tested for multiple times without error.

matthew-brett committed:
BUG: deprecation for ignored corrcoef args

The `ddof` and `bias` arguments to `corrcoef` cancel in the math of the
correlation coefficient, so there is no point in passing these values,
and it is confusing to the user to see these in the docstring.

Remove, deprecate, and test.

matthew-brett committed:
ENH+TST: add context manager for testing warnings

Warnings can be slippery, because, whenever a warning is
triggered, Python adds a __warningregistry__ member to the *calling*
module.  This makes it impossible to retrigger the warning in this
module, whatever you put in simplefilter.  The `catch_warn_reset`
decorator removes the __warningregistry__ member as the context
manager exits, making it possible to retrigger the warning.

njsmith commented:
You mean "context manager", not " decorator". Everything else looks good to
me (on a quick skim at least).
On Mar 12, 2015 5:32 PM, "Matthew Brett" <notifications@github.com> wrote:

> As discussed on the mailing list, the values for the bias and ddof
> arguments to the corrcoef cancel in calculation, and are therefore
> pointless and confusing.
>
> Deprecate these arguments and document their pending removal.
>
> This PR also includes a version of the warnings.catch_warnings decorator
> that allows warnings to be tested for multiple times without error.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/numpy/numpy/pull/5675
> Commit Summary
>
>    - ENH+TST: add decorator for testing warnings
>    - BUG: deprecation for ignored corrcoef args
>
> File Changes
>
>    - *M* numpy/lib/function_base.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-0> (45)
>    - *M* numpy/lib/tests/test_function_base.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-1> (31)
>    - *M* numpy/ma/extras.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-2> (50)
>    - *M* numpy/ma/tests/test_extras.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-3> (76)
>    - *M* numpy/ma/tests/test_regression.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-4> (15)
>    - *A* numpy/testing/tests/test_warnutils.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-5> (48)
>    - *A* numpy/testing/warnutils.py
>    <https://github.com/numpy/numpy/pull/5675/files#diff-6> (36)
>
> Patch Links:
>
>    - https://github.com/numpy/numpy/pull/5675.patch
>    - https://github.com/numpy/numpy/pull/5675.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5675>.
>

matthew-brett commented:
Thanks - I fixed references to 'decorator'.

The main question that I have is what to do about the 'bias' argument to `np.ma.extras.corrcoef`.   It is before a still-used argument `allow_masked`.

It might be prudent to plan to make `allow_masked` a keyword-only argument at the same time as the end of the deprecation period for `bias`, so intended inputs to `bias` don't get picked up by `allow_masked`.  This would involve deprecating `allow_masked` as a positional argument.

jaimefrio commented on numpy/lib/function_base.py:
```diff
@@ -1975,28 +1984,28 @@ def corrcoef(x, y=None, rowvar=1, bias=0, ddof=None):
+    arguments had no effect on the return values of the function and can be
+    safely ignored in this and previous versions of numpy.
+    """
+    if len(args) or kwargs.pop('bias', _DefaultArg) is not _DefaultArg:
```
We are opening the door to lots of nonsense input going through silently, e.g. extra keyword or positional arguments, `bias` and `ddof` being provided as both positional and keyword simultaneously... May not be worth adding all those checks for something we actually want to deprecate, but...

matthew-brett replied:
Checking for simultaneous positional and keyword does seem too much for arguments that don't have any effect on the function.  We are already warning about their use.

It's easy to check for the length of args and kwargs after parsing though.

matthew-brett committed:
ENH: make ma.corrcoef allow_masked keyword-only

"allow_masked" follows "bias" argument, but "bias" will go away at some
point.  Intend that we will make "allow_masked" keyword only at the same
time that we remove "bias".

matthew-brett committed:
ENH: raise errors for wrong arguments in corrcoef

Raise errors for the wrong number of positional arguments or incorrect
keyword arguments in corrcoef routines.

matthew-brett commented:
I added a deprecation warning for `allow_masked` as positional argument as above.

I added checks for number of positional args and unexpected keyword args.

charris commented on numpy/ma/extras.py:
```diff
@@ -48,7 +48,7 @@
 import numpy.core.umath as umath
 from numpy.lib.index_tricks import AxisConcatenator
-from numpy.linalg import lstsq
+from numpy.lib.function_base import _CORRCOEF_MSG_FMT, _DefaultArg
```
Just repeat the definition here. This import complicates the code, obscures the value of _CORRCOEF_MSG_FMT, and introduces a dependency between modules. The format is temporary anyway. No reason not to also put the private `_DefaultArg` here also, and for the same reasons.

_DefaultArg probably belongs somewhere else where it can be reused.

matthew-brett replied:
Willco.

Any suggestions for a good place for `_DefaultArg`?

charris replied:
Since its use is entirely local, it might be better to just define it in each module in which it is used. It is going to be temporary, at least in the numpy sense of temporary. Otherwise, I might suggest `numpy/__init__.py`, which already has a couple of global classes. In the latter case a more suggestive name might help.

matthew-brett committed:
ENH: replicate code support fragments for clarity

Rather than importing the small code support fragments from
function_base.py, replicate them to make the code easier to read.

matthew-brett committed:
DOC: update docstring for ma.extras.corrcoef

Note that this is the Pearson product-moment correlation.

Note the pending change of allow_masked to keyword-only.

matthew-brett committed:
ENH: allow default modules for catch_warn_reset

Add ability to add default modules to catch_warn_reset classs, by
inheritance.

charris commented:
Both the deprecation and context manager should be mentioned in the release notes. Apropos the context manager,  a separate PR might be best. A name was closer to the standard `catch_warnings` would also be good something like `catch_and_reset_warnings`. I also wonder if it wouldn't be possible to just integrate it into `assert_warns`?

charris commented:
Or maybe `catch_and_clear_warnings` for added alliteration.

matthew-brett committed:
ENH: use new catch_warn_reset inheritance

Use inheritance of catch_warn_reset for slightly cleaner context
managers.

matthew-brett commented:
No problem for separate PR, and renaming of context manager.

Adding to `assert_warns` would need a new 'modules' argument to `assert_warns`, at least the way I have written it.

Sckit-image go for a more magic and more nuclear solution, always clearing out all recorded previous warnings for modules in the call stack:

https://github.com/scikit-image/scikit-image/blob/master/skimage/_shared/_warnings.py

charris commented:
Google also turned up that nuclear option. So there is no easy way to know what module a warning was raised in when it is caught?

matthew-brett commented:
I don't know of any way to find out what module the warning was raised in, from the warning itself...

charris commented:
Thinking about it, both the format and class could be local to the function, that way everything could be deleted at one go when it comes to that.

matthew-brett commented:
My worry about putting the class definition and format string in the funciton was that it made the meat of the function harder to read.  I can imagine that it might be distracting to see this little piece of uncecessarily repeated work inside the function.

charris commented:
Good point. OTOH, if it is outside the function one wonders where else it is used.

matthew-brett committed:
ENH: move format strings into body of function

Move the format strings into the function, as they are only used in the
function.

matthew-brett commented:
I was hoping that wondering where else it is used it more typical of the keen-eyed maintainer than the casual developer or user.

I've moved just the string into the function - how about that?

charris commented:
OK. Too bad we can't just use `None`. Hmm..., if the default value of `ddof` is made 0, I think `None` would be OK.

matthew-brett commented:
My worry with None is that it's not possible to tell whether the user passed it, thinking the argument still existed.

charris commented:
How about just check for the key, then pop it if present and issue the warning?

matthew-brett commented:
Yes, I could split the check, for each of bias and ddof, into:

1) Has it been passed as a positional argument?  Then warn
2) Has it been passed as a keyword argument? Then warn

So - something like this for each argument:

```
    if nargs:
        warnings.warn(arg_msg_fmt.format('bias'), DeprecationWarning)
    if 'bias' in kwargs:
        kwargs.pop('bias')
        warnings.warn(arg_msg_fmt.format('bias'), DeprecationWarning)
```

It is clearer, just longer - would you prefer that?

charris commented:
Yes, that looks better and we don't need to worry about the special class.

matthew-brett committed:
ENH: simpler checks of corrcoef args and kwargs

Drop fancy class and do it the long way to make the checks more
readable.

matthew-brett commented:
Like so?

charris commented:
Looks better ;) I think you can keep the combination of narg check and key check, just pop with default None.

matthew-brett commented:
The problem with popping with default None is that the check will fail (no warning raised) when the user passes None as the value of the keyword argument, so for example the check:

```
    if nargs or kwargs.pop('bias', None) is not None:
```

will evaluate to False if the user passes ``corrcoef(x, bias=None)``.

That was the problem I was trying to solve with the fancy class...

charris commented:
But you have already checked for the keyword. I mean something like
```
if nargs or 'bias' in kwargs:
    kwargs.pop('bias', None)
    warnings.warn(arg_msg_fmt.format('bias'), DeprecationWarning)
```

matthew-brett commented:
Ah - sorry - I get what you mean.

charris commented on numpy/lib/function_base.py:
```diff
@@ -1975,28 +1975,40 @@ def corrcoef(x, y=None, rowvar=1, bias=0, ddof=None):
+    if nargs > 2:
+        raise TypeError(
+            'corrcoef takes at most 5 arguments ({0} given)'.format(nargs))
+    arg_msg_fmt = ("The {0} argument has no effect on the result of corrcoef "
```
Could probably use a `\n` to break the error message.

charris commented:
OK, that looks good. I'd probably just do `fmt = `, but that is a nitpick. Do you want to make a PR for the (renamed) context manager? I can put it in, then you can fix things up here and remove its definition from here.

charris commented:
Don't worry about the test, looks like the waf connection is failing.

matthew-brett committed:
ENH: add line break in corrcoef error message

matthew-brett committed:
ENH: rename format argument

matthew-brett commented:
Your nit-pick is my command.

If this is all good I'll rebase into two PRs.

charris commented on numpy/lib/function_base.py:
```diff
@@ -1975,28 +1975,40 @@ def corrcoef(x, y=None, rowvar=1, bias=0, ddof=None):
+        warnings.warn(fmt.format('ddof'), DeprecationWarning)
+    if len(kwargs):
+        raise TypeError(
+            "corrcoef got an unexpected keyword argument '{0}'".format(
```
Note that you can break this line
```
raise TypeError("corrcoef got an unexpected keyword "
                "argument '{0}'".format(list(kwargs)[0])

charris commented:
LGTM, go for it. I tossed in another nitpick...

matthew-brett committed:
ENH: rename catch_warn_reset to catch_clear_warnings

matthew-brett commented:
OK to rename context manager to `catch_clear_warnings` ?

matthew-brett committed:
ENH: reformat another warning string

charris commented:
I'd probably leave the `_and_` in there.

charris commented:
Although I confess to being a bit unsure why the context manager is needed. Does this fix a problem when the tests are run in interactive mode?

matthew-brett committed:
ENH: move warnings context manager, rename

matthew-brett commented:
I moved the context manager into `numpy.testing.utils` and added more docstring - does the docstring help explain the problem the context manager is trying to solve?

matthew-brett committed:
DOC: fix trailing rename

matthew-brett committed:
ENH: reformat multiple imports

matthew-brett commented:
OK to go ahead with 2 PRs from this state?

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -1,10 +1,16 @@
 from __future__ import division, absolute_import, print_function
 
 import warnings
+from warnings import warn, simplefilter
```
We usually just import warnings and use it in the `warnings.warn` form.

charris commented on numpy/testing/utils.py:
```diff
@@ -1718,3 +1718,44 @@ def tempdir(*args, **kwargs):
+    * resets ``__warningregistry__`` to its previous state on exit.
+
+    This makes it possible to trigger any warning afresh inside the context
+    manager without disturbing the state of warnings outside.
```
Need to document the arguments for the constructor here. An example showing how to get the module name would be good.

matthew-brett committed:
ENH: use 'warnings' module instead of importing from

charris commented on numpy/testing/utils.py:
```diff
@@ -1718,3 +1718,44 @@ def tempdir(*args, **kwargs):
+    """
+    class_modules = ()
+
+    def __init__(self, *args, **kwargs):
```
We discourage the use of `*args` and `**kwargs`, better to lay out the actual args. Looks like the only other ones you need are `modules` and `record`.

charris commented:
Sure. I won't guarantee no more comments, but I think this is getting close.

matthew-brett committed:
DOC: fill in docstring, make args, kwargs explicit

matthew-brett commented:
Maybe the context manager should better be called 'catch_all_warnings_mods' or something else to reflect the fact that it only resets the warnings registry inside the context manager.

matthew-brett commented:
Or `clear_and_catch_warnings`

matthew-brett committed:
ENH: rename warnings context manager

matthew-brett commented:
OK - I think this is ready now - will split into separate PRs soon.

matthew-brett commented:
Context manager PR here : https://github.com/numpy/numpy/pull/5682

matthew-brett commented:
Deprecation stuff here : https://github.com/numpy/numpy/pull/5683

charris commented:
OK, I'll close this now.
